File: Signal/src/ViewControllers/ThreadSettings/OWSAddToContactViewController.m
//
//  Copyright (c) 2019 Open Whisper Systems. All rights reserved.
//

#import "OWSAddToContactViewController.h"
#import <ContactsUI/ContactsUI.h>
#import <SignalMessaging/ContactsViewHelper.h>
#import <SignalMessaging/Environment.h>
#import <SignalMessaging/OWSContactsManager.h>
#import <SignalMessaging/UIUtil.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@interface OWSAddToContactViewController () <CNContactViewControllerDelegate, ContactsViewHelperDelegate>

@property (nonatomic) SignalServiceAddress *address;

@property (nonatomic, readonly) OWSContactsManager *contactsManager;
@property (nonatomic, readonly) ContactsViewHelper *contactsViewHelper;

@end

#pragma mark -

@implementation OWSAddToContactViewController

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return self;
    }

    [self commonInit];

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (!self) {
        return self;
    }

    [self commonInit];

    return self;
}

- (instancetype)initWithNibName:(nullable NSString *)nibNameOrNil bundle:(nullable NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (!self) {
        return self;
    }

    [self commonInit];

    return self;
}

- (void)commonInit
{
    _contactsManager = Environment.shared.contactsManager;
    _contactsViewHelper = [[ContactsViewHelper alloc] initWithDelegate:self];
}

- (void)configureWithAddress:(SignalServiceAddress *)address
{
    OWSAssertDebug(address.isValid);

    _address = address;
}

#pragma mark - CNContactViewControllerDelegate

- (void)contactViewController:(CNContactViewController *)viewController
       didCompleteWithContact:(nullable CNContact *)contact
{
    OWSAssertDebug(self.navigationController);

    // We're done, pop back to the view that presented us.

    NSUInteger selfIndex = [self.navigationController.viewControllers indexOfObject:self];
    if (selfIndex == NSNotFound) {
        OWSFailDebug(@"Unexpectedly not in nav hierarchy");
        return;
    }

    UIViewController *previousVC = self.navigationController.viewControllers[selfIndex - 1];

    [self.navigationController popToViewController:previousVC animated:YES];
}

#pragma mark - ContactsViewHelperDelegate

- (void)contactsViewHelperDidUpdateContacts
{
    [self updateTableContents];
}

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];

    self.title = NSLocalizedString(@"CONVERSATION_SETTINGS_ADD_TO_EXISTING_CONTACT",
        @"Label for 'new contact' button in conversation settings view.");

    [self updateTableContents];
}

- (nullable NSString *)displayNameForContact:(Contact *)contact
{
    OWSAssertDebug(contact);

    if (contact.fullName.length > 0) {
        return contact.fullName;
    }

    for (NSString *email in contact.emails) {
        if (email.length > 0) {
            return email;
        }
    }
    for (NSString *phoneNumber in contact.userTextPhoneNumbers) {
        if (phoneNumber.length > 0) {
            return phoneNumber;
        }
    }

    return nil;
}

- (void)updateTableContents
{
    OWSTableContents *contents = [OWSTableContents new];
    contents.title = NSLocalizedString(@"CONVERSATION_SETTINGS", @"title for conversation settings screen");

    __weak OWSAddToContactViewController *weakSelf = self;

    OWSTableSection *section = [OWSTableSection new];
    section.headerTitle = NSLocalizedString(
        @"EDIT_GROUP_CONTACTS_SECTION_TITLE", @"a title for the contacts section of the 'new/update group' view.");

    for (Contact *contact in self.contactsViewHelper.contactsManager.allContacts) {
        NSString *_Nullable displayName = [self displayNameForContact:contact];
        if (displayName.length < 1) {
            continue;
        }

        // TODO: Confirm with nancy if this will work.
        NSString *cellName = [NSString stringWithFormat:@"contact.%@", NSUUID.UUID.UUIDString];
        [section addItem:[OWSTableItem disclosureItemWithText:displayName
                                      accessibilityIdentifier:ACCESSIBILITY_IDENTIFIER_WITH_NAME(self, cellName)
                                                  actionBlock:^{
                                                      [weakSelf presentContactViewControllerForContact:contact];
                                                  }]];
    }
    [contents addSection:section];

    self.contents = contents;
    [self.tableView reloadData];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
}

#pragma mark - Actions

- (void)presentContactViewControllerForContact:(Contact *)contact
{
    OWSAssertDebug(contact);
    OWSAssertDebug(self.address.isValid);

    if (!self.contactsManager.supportsContactEditing) {
        OWSFailDebug(@"Contact editing not supported");
        return;
    }
    CNContact *_Nullable cnContact;
    if (contact.cnContactId != nil) {
        cnContact = [self.contactsManager cnContactWithId:contact.cnContactId];
        OWSAssertDebug(cnContact != nil);
    }
    CNContactViewController *_Nullable contactViewController =
        [self.contactsViewHelper contactViewControllerForAddress:self.address
                                                 editImmediately:YES
                                          addToExistingCnContact:cnContact];

    if (!contactViewController) {
        OWSFailDebug(@"Unexpected missing contact VC");
        return;
    }

    contactViewController.delegate = self;

    [self.navigationController pushViewController:contactViewController animated:YES];
}

@end

NS_ASSUME_NONNULL_END

File: Signal/src/ViewControllers/ThreadSettings/OWSAddToContactViewController.swift
//
// Copyright 2020 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Contacts
import ContactsUI
import SignalServiceKit
import SignalMessaging
import SignalUI

public class OWSAddToContactViewController: OWSViewController {

    private let address: SignalServiceAddress
    private var tableView: UITableView!

    private lazy var contacts = [Contact]()

    private let sortOrder: CNContactSortOrder = CNContactsUserDefaults.shared().sortOrder

    public init(address: SignalServiceAddress) {
        self.address = address
        super.init()
        contactsViewHelper.addObserver(self)
    }

    public override func loadView() {
        view = UIView()

        tableView = UITableView()

        view.addSubview(tableView)
        tableView.autoPinEdge(toSuperviewEdge: .top)
        tableView.autoPinEdge(toSuperviewEdge: .bottom)
        tableView.autoPinEdge(toSuperviewSafeArea: .leading)
        tableView.autoPinEdge(toSuperviewSafeArea: .trailing)
        tableView.delegate = self
        tableView.dataSource = self
    }

    public override func viewDidLoad() {
        super.viewDidLoad()

        // Auto size cells for dynamic type
        tableView.estimatedRowHeight = 60.0
        tableView.rowHeight = UITableView.automaticDimension
        tableView.estimatedRowHeight = 60

        tableView.allowsMultipleSelection = false

        tableView.register(ContactCell.self, forCellReuseIdentifier: ContactCell.reuseIdentifier)

        title = OWSLocalizedString(
            "CONVERSATION_SETTINGS_ADD_TO_EXISTING_CONTACT",
            comment: "Label for 'new contact' button in conversation settings view."
        )

        updateData()
        tableView.reloadData()
    }

    public override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)

        view.backgroundColor = Theme.backgroundColor
        tableView.backgroundColor = Theme.backgroundColor
        tableView.separatorColor = Theme.cellSeparatorColor
    }

    fileprivate func presentContactViewController(forContact contact: Contact) {
        guard let navigationController else {
            return owsFailDebug("Missing navigationController.")
        }
        contactsViewHelper.checkEditingAuthorization(
            authorizedBehavior: .pushViewController(on: navigationController, viewController: {
                let cnContact = self.contactsManager.cnContact(withId: contact.cnContactId)
                let result = self.contactsViewHelper.contactViewController(
                    for: self.address,
                    editImmediately: true,
                    addToExisting: cnContact
                )
                result.delegate = self
                return result
            }),
            unauthorizedBehavior: .presentError(from: self)
        )
    }

    private func updateData() {
        contacts = databaseStorage.read { transaction in
            contactsManagerImpl.allSortedContacts(transaction: transaction)
        }
    }
}

extension OWSAddToContactViewController: UITableViewDelegate {

    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        presentContactViewController(forContact: contacts[indexPath.row])
    }
}

extension OWSAddToContactViewController: UITableViewDataSource {

    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        contacts.count
    }

    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let contact = contacts[safe: indexPath.row] else {
            owsFailDebug("failed to lookup contact")
            return UITableViewCell()
        }

        let cell = tableView.dequeueReusableCell(ContactCell.self, for: indexPath)!
        cell.configure(contact: contact, sortOrder: sortOrder, subtitleType: .none, showsWhenSelected: false)
        return cell
    }

}

extension OWSAddToContactViewController: ContactsViewHelperObserver {
    public func contactsViewHelperDidUpdateContacts() {
        updateData()
        tableView.reloadData()
    }
}

extension OWSAddToContactViewController: CNContactViewControllerDelegate {

    public func contactViewController(_ viewController: CNContactViewController, didCompleteWith contact: CNContact?) {
        guard let index = navigationController?.viewControllers.firstIndex(of: self), index > 0,
            let previousVC = (navigationController?.viewControllers[index - 1]) else { return }

        navigationController?.popToViewController(previousVC, animated: true)
    }

}
